resource-aggregate/events/servicesMetadataSnapshotTaken.go

```diff
@@ -0,0 +1,328 @@
+		case (&ServicesMetadataSnapshotTaken{}).EventType():
+			var s ServicesMetadataSnapshotTaken
+			if err := eu.Unmarshal(&s); err != nil {
+				return status.Errorf(codes.Internal, "%v", err)
+			}
+			d.HandleServicesMetadataSnapshotTaken(ctx, &s)
+		case (&ServicesMetadataUpdated{}).EventType():
+			var s ServicesMetadataUpdated
+			if err := eu.Unmarshal(&s); err != nil {
+				return status.Errorf(codes.Internal, "%v", err)
+			}
+			_, _ = d.HandleServicesMetadataUpdated(ctx, &s)
+		}
+	}
+	return iter.Err()
+}
+
+func serviceStatusMapToArray(m map[string]*ServicesStatus_Status) []*ServicesStatus_Status {
+	arr := make([]*ServicesStatus_Status, 0, len(m))
+	for _, v := range m {
+		arr = append(arr, v)
+	}
+	// sort by serviceId and instanceId to make sure that order of services is always the same
+	sort.Slice(arr, func(i, j int) bool {
+		return arr[i].GetId() < arr[j].GetId()
+	})
+	return arr
+}
+
+func (d *ServicesMetadataSnapshotTaken) updateStatus(ctx context.Context, req *commands.UpdateServiceMetadataRequest, em *EventMetadata, ac *commands.AuditContext) ([]eventstore.Event, error) {
+	if req.GetStatus().GetId() == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "cannot update status for device %v: invalid instanceId", req.GetStatus().GetId())
+	}
+	now := time.Now()
+	servicesStatus := d.GetServicesMetadataUpdated().GetStatus()
+	offline := make(map[string]*ServicesStatus_Status, len(servicesStatus.GetOnline())+len(servicesStatus.GetOffline()))
+	for _, v := range servicesStatus.GetOffline() {
+		key := v.GetId()
+		offline[key] = v
+	}
+
+	if offline[req.GetStatus().GetId()] != nil {
+		// The service is already offline, and the service needs to be shut down to avoid conflicts in device connection status (ONLINE/OFFLINE).
+		return nil, status.Errorf(codes.FailedPrecondition, "cannot update status for device %v: already offline", req.GetStatus().GetId())
+	}
+
+	online := make(map[string]*ServicesStatus_Status, 1)
+	for _, v := range servicesStatus.GetOnline() {
+		key := v.GetId()
+		if v.GetOnlineValidUntil() < now.UnixNano() {
+			offline[key] = v
+		}
+	}
+	key := req.GetStatus().GetId()
+
+	// When the processing time exceeds the timeToLive, we must extend the timeToLive.
+	// During processing, several tasks are performed: receiving requests, loading data from the database, and processing requests.
+	// Writing to the database is an expensive operation, so if the processing time is more than 10 times the timeToLive, we should extend it.
+	// During idle times, the processing time is typically around 10 milliseconds.
+	// Under heavy load, the processing time can increase to approximately 5 minutes.
+	timeToLive := time.Duration(req.GetStatus().GetTimeToLive())
+	if req.GetStatus().GetTimestamp() >= 0 {
+		processingTime := time.Since(pkgTime.Unix(0, req.GetStatus().GetTimestamp()))
+		if processingTime*writeCostAgainstRead > timeToLive {
+			timeToLive = processingTime * writeCostAgainstRead
```

Comment:
So under heavy load this could be 50+ minutes, does the next metadata update override the expiration time?

Reply:
Yes it will be updated.